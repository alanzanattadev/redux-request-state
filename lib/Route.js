"use babel";
// @flow

import type { ResolveResult } from "./types.js";
import type { ResolveFn } from "./types.js";
import type { ReplayWithFn } from "./types.js";
import type { RequestIDFactory } from "./types.js";
import type { RequestAction } from "./types.js";

export default class Route {
  type: string;
  resolve: ResolveFn;
  replayWith: ?ReplayWithFn;
  requestIDFactory: RequestIDFactory;
  constructor(
    {
      type = "REQUEST",
      resolve = () => Promise.resolve(null),
      replayWith = null,
      requestIDFactory,
    }: {
      type: string,
      resolve: ResolveFn,
      replayWith: ?ReplayWithFn,
      requestIDFactory: RequestIDFactory,
    } = {},
  ) {
    if (typeof requestIDFactory !== "function")
      throw new Error(
        "You have to define a requestIDFactory attribut to the Route constructor: (...params) => string",
      );
    this.type = type;
    this.resolve = resolve;
    this.replayWith = replayWith;
    this.requestIDFactory = requestIDFactory;
  }

  getRequestID(...params: Array<any>): string {
    return this.requestIDFactory(...params);
  }

  create(...params: Array<any>): RequestAction {
    return {
      type: this.type,
      resolve: () => this.resolve(...params),
      replayWith: this.replayWith,
      requestID: this.requestIDFactory(...params),
    };
  }

  createAndInject(resolver: ResolveResult, ...params: Array<any>): RequestAction {
    return {
      type: this.type,
      resolve: () => resolver,
      replayWith: this.replayWith,
      requestID: this.requestIDFactory(...params),
    };
  }
}
