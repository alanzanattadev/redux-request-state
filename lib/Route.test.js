"use babel";
// @flow

import Route from "./Route";

describe("Route", () => {
  it("should get request ID correctly", () => {
    let subject = new Route({
      requestIDFactory: (user, photoID) => `user.${user.id}.photo.${photoID}`,
    });

    expect(subject.getRequestID({ id: "2" }, "myphoto")).toBe(
      "user.2.photo.myphoto",
    );
  });

  it("should create action correctly", (done) => {
    let request = new Route({
      requestIDFactory: (user, photoID) => `user.${user.id}.photo.${photoID}`,
      type: "MY_REQUEST",
      resolve: (user, photoID) =>
        Promise.resolve({ id: user.id, photoID: photoID }),
      replayWith: response => ({ type: "PONG" }),
    });
    let subject = request.create({ id: "1" }, "myphoto");

    expect(subject).toMatchSnapshot();
    subject.resolve().then((response) => {
      expect(response).toMatchSnapshot();
      done();
    })
  });
});
