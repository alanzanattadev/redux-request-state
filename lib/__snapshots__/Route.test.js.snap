// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Route should create action correctly 1`] = `
Object {
  "replayWith": [Function],
  "requestID": "user.1.photo.myphoto",
  "resolve": [Function],
  "type": "MY_REQUEST",
}
`;

exports[`Route should create action correctly 2`] = `
Object {
  "id": "1",
  "photoID": "myphoto",
}
`;
